File: domain/commands/handlers.py
from typing import Any, Callable, Optional

from domain.models.command_result import CommandResult
from models.config import LLMConfig
from utils.logger import Logger

CommandHandler = Callable[..., CommandResult]


def handle_exit(**kwargs: Any) -> CommandResult:
    """
    Handle the /exit command to terminate the Xabiro agent session.
    """
    return CommandResult(
        message="Até mais! Xabiro encerrado.",
        should_exit=True,
    )


def handle_help(**kwargs: Any) -> CommandResult:
    """
    Handle the /help command to provide a list of available commands.
    """
    help_message = (
        "Comandos disponíveis:\n"
        "/help - Mostrar esta mensagem de ajuda\n"
        "/exit - Sair do Xabiro\n"
        "/config - Mostra a Config atual do Xabiro\n"
    )
    return CommandResult(message=help_message, should_exit=False)


def handle_toggle_logging(**kwargs: Any) -> CommandResult:
    """
    Handle the /toggle_logging command to enable or disable logging.
    """
    logger: Optional[Logger] = kwargs.get("logger")

    if logger is None:
        raise ValueError(
            "Dependência 'logger' é obrigatória para o comando /toggle_logging."
        )

    logger.enabled = not logger.enabled
    status = "ativado" if logger.enabled else "desativado"
    return CommandResult(message=f"Logging {status}.", should_exit=False)


def handle_config(**kwargs: Any) -> CommandResult:
    """
    Handle the /config command to return the current status of the Xabiro agent.
    """
    llm_config: Optional[LLMConfig] = kwargs.get("llm_config")

    if llm_config is None:
        raise ValueError(
            "Dependência 'llm_config' é obrigatória para o comando /status."
        )

    status_msg = (
        f"Config atual do Xabiro:\n"
        f"  Modelo: {llm_config.model}\n"
        f"  Temperatura: {llm_config.temperature}\n"
        f"  Max Tokens: {llm_config.max_tokens}"
    )
    return CommandResult(message=status_msg, should_exit=False)

File: src/domain/commands/handlers.ts
/**
 * Command handlers for the CLI commands.
 */

import * as path from 'path';
import type { CommandResult } from '../models/command.js';
import { createCommandResult } from '../models/command.js';
import type { Logger } from '../../utils/logger.js';
import type { LLMConfig } from '../models/config.js';
import type { WorkingDirectoryManager } from '../pattern/working-directory-manager.js';

export type CommandHandler = (args: CommandHandlerArgs) => CommandResult;

export interface CommandHandlerArgs {
  commandArguments?: string;
  logger?: Logger;
  llmConfig?: LLMConfig;
  workingDirectoryManager?: WorkingDirectoryManager;
}

/**
 * Handle the /cd command to change the working directory.
 */
export function handleChangeWorkingDirectory(args: CommandHandlerArgs): CommandResult {
  const { commandArguments, workingDirectoryManager } = args;

  if (!workingDirectoryManager) {
    throw new Error("Dependency 'workingDirectoryManager' is required for the /cd command.");
  }

  if (!commandArguments) {
    return createCommandResult(
      "Erro: O argumento 'new_directory' é obrigatório para o comando /cd.",
      false
    );
  }

  try {
    const absDirectory = path.resolve(commandArguments);
    workingDirectoryManager.setDirectory(absDirectory);
    return createCommandResult(
      `Diretório de trabalho alterado para: ${absDirectory}`,
      false
    );
  } catch (error) {
    return createCommandResult((error as Error).message, false);
  }
}

/**
 * Handle the /exit command to terminate the session.
 */
export function handleExit(_args: CommandHandlerArgs): CommandResult {
  return createCommandResult("Até mais! Xabiro encerrado.", true);
}

/**
 * Handle the /help command to provide available commands.
 */
export function handleHelp(_args: CommandHandlerArgs): CommandResult {
  const helpMessage = `Comandos disponíveis:
/help - Mostrar esta mensagem de ajuda
/exit - Sair do Xabiro
/config - Mostra a Config atual do Xabiro
/toggle_logging - Ativar/Desativar logging
/cd <diretório> - Alterar o diretório de trabalho do Xabiro`;

  return createCommandResult(helpMessage, false);
}

/**
 * Handle the /toggle_logging command to enable or disable logging.
 */
export function handleToggleLogging(args: CommandHandlerArgs): CommandResult {
  const { logger } = args;

  if (!logger) {
    throw new Error("Dependency 'logger' is required for the /toggle_logging command.");
  }

  logger.enabled = !logger.enabled;
  const status = logger.enabled ? "ativado" : "desativado";
  return createCommandResult(`Logging ${status}.`, false);
}

/**
 * Handle the /config command to return the current configuration.
 */
export function handleConfig(args: CommandHandlerArgs): CommandResult {
  const { llmConfig } = args;

  if (!llmConfig) {
    throw new Error("Dependency 'llmConfig' is required for the /config command.");
  }

  const statusMsg = `Config atual do Xabiro:
  Modelo: ${llmConfig.model}
  Temperatura: ${llmConfig.temperature}
  Max Tokens: ${llmConfig.max_tokens}`;

  return createCommandResult(statusMsg, false);
}
